File: LeetCode/Graph/DFS/Golang/Leetcode558-logical-or-of-two-binary-grids-represented-as-quad-trees.go
package sbgo

func intersect(t1 *Node, t2 *Node) *Node {
    if t1.IsLeaf {
        if t1.Val {
            return &Node{Val:true,IsLeaf:true}
        }
        return &Node{t2.Val,t2.IsLeaf,t2.TopLeft,t2.TopRight,t2.BottomLeft,t2.BottomRight}
    }
    if t2.IsLeaf {
        return intersect(t2,t1);
    }
    o1 := intersect(t1.TopLeft,t2.TopLeft)
    o2 := intersect(t1.TopRight,t2.TopRight)
    o3 := intersect(t1.BottomLeft,t2.BottomLeft)
    o4 := intersect(t1.BottomRight,t2.BottomRight);
    if o1.IsLeaf && o2.IsLeaf && o3.IsLeaf && o4.IsLeaf && o1.Val == o2.Val && o1.Val == o3.Val && o1.Val == o4.Val {
        return &Node{Val:o1.Val,IsLeaf:true}
    }
    return &Node{false,false,o1,o2,o3,o4}
}

File: LeetCode/Graph/DFS/JS/Leetcode558-logical-or-of-two-binary-grids-represented-as-quad-trees.js
var intersect = function(t1, t2) {
    if(t1.isLeaf) {
        if(t1.val) {
            return new Node(true,true);
        }
        return new Node(t2.val,t2.isLeaf,t2.topLeft,t2.topRight,t2.bottomLeft,t2.bottomRight);
    }
    if(t2.isLeaf) {
        return intersect(t2,t1);
    }
    let o1 = new intersect(t1.topLeft,t2.topLeft);
    let o2 = new intersect(t1.topRight,t2.topRight);
    let o3 = new intersect(t1.bottomLeft,t2.bottomLeft);
    let o4 = new intersect(t1.bottomRight,t2.bottomRight);
    if(o1.isLeaf && o2.isLeaf && o3.isLeaf && o4.isLeaf && o1.val == o2.val && o1.val == o3.val && o1.val == o4.val) {
        return new Node(o1.val,true);
    }
    return new Node(false,false,o1,o2,o3,o4);
};
